### PR TITLE
fix(runners): prevent orphaned MCP processes via parent-liveness watcher

### LINKED ISSUE #56: bug: orphaned bun processes accumulate across Claude Code sessions
## Summary

MCP runner processes (bun-runner, biome-runner, tsc-runner, x-api) are not being cleaned up when Claude Code sessions end. Over time this leads to dozens of orphaned `bun` processes consuming system resources.

## Observed behavior

After a day of work with multiple Claude Code sessions, `pgrep -la bun` showed **39 bun processes** -- only 8 belonged to the active session. The rest were orphans from terminated sessions whose parent Claude processes were already gone.

## Expected behavior

When a Claude Code session ends (or its parent process dies), child MCP server processes should be cleaned up. Bun processes should not outlive their parent session.

## Details

- Each Claude Code session spawns ~4 bun MCP servers (bun-runner, biome-runner, tsc-runner, x-api)
- When the Claude process exits, the bun children become orphaned (reparented to PID 1)
- These orphans persist indefinitely until manually killed
- With multiple sessions per day (and worktree sessions), this compounds quickly

## Environment

- macOS Tahoe 26.2, Apple Silicon
- Bun v1.x
- Claude Code with `--plugin-dir` loading

## Possible fixes

- Register a `SIGTERM`/`SIGHUP` handler in each MCP server that exits gracefully when the parent dies
- Use process groups so killing the parent cascades to children
- Add a periodic self-check: if `process.ppid === 1`, exit (parent is gone)
- Add a `Stop` lifecycle hook that explicitly kills child MCP processes

### PR DESCRIPTION
**Pull request by nathanvale**
## Summary

Adds a defensive parent-liveness watcher to all three MCP runners (\`bun-runner\`, \`biome-runner\`, \`tsc-runner\`) so they exit promptly when their parent Claude Code session (or sub-agent) dies without delivering SIGTERM. Closes #56.

Per the issue: each Claude Code session spawns ~4 bun MCP servers, and every sub-agent invocation spawns 4 more. When the parent dies abruptly, runners get reparented to PID 1 and persist indefinitely — a typical workday accumulates 30+ orphaned bun processes.

The watcher polls \`process.ppid\` every 5s (configurable via \`MCP_PARENT_CHECK_MS\`) and routes through the existing \`shutdown()\` path when divergence or PID 1 is detected. It layers additively on top of \`transport.onclose\` + SIGINT/SIGTERM — those primary paths are unchanged. The interval handle is \`unref()\`'d so it never keeps the loop alive on its own.

## What changed

- **3 runner packages** — added \`DEFAULT_PARENT_CHECK_MS\`, \`MIN_PARENT_CHECK_MS\`, \`parseParentCheckMs()\`, \`createParentLivenessWatcher()\`; wired into each \`start*Server()\` after the existing signal handlers.
- **3 test files** — 13 unit tests per package covering env parsing edge cases (default, empty, NaN, zero, negative, clamping) and watcher behavior (disabled / unref / no-fire when ppid unchanged / fire on change / fire on PID 1).
- **\`scripts/smoke/run-smoke.ts\`** — new \`orphan-detection\` smoke case loops over all 3 built \`dist/index.js\` binaries × 3 scenarios (happy path 200ms / negative control disabled / edge case 10s). Negative control proves the watcher (not \`transport.onclose\`) is the active mechanism.
- **Plan doc** — status flipped \`active\` → \`completed\`.
- **Changeset** — patch bump for all three runners.

The watcher logic is duplicated across the three packages by design — repo policy keeps each runner self-contained (\`.claude/CLAUDE.md\`).

## Validation

\`bun run validate\` is green: 142 unit tests + 5 smoke cases. Orphan-detection smoke completes in 10.1s and includes a negative-control gate that fails loudly if stdin-detachment is leaking (which would mask whether the watcher is the actual mechanism).

Manual sanity: built runner binary, killed its parent shell, confirmed exit within ~5s.

## Test plan

- [x] \`bun run validate\` green locally
- [x] Manual: confirm built runner exits within ~5s of parent death
- [ ] CI green (\`pr-quality\`, \`commitlint\`, \`workflow-lint\`, \`package-hygiene\`, \`node-compat\`)
- [ ] Behavior verification on a Claude Code session: orphan count over a workday with sub-agent activity

## Code review notes (deferred follow-ups)

The \`/ce-code-review\` pass surfaced several robustness concerns that were investigated but **deliberately not applied** to this PR. The original watcher works empirically (smoke proves it), but the macOS Bun \`process.ppid\` semantics are subtler than my fixes assumed. Filed as follow-ups for separate diagnosis:

- **\`initialPpid === 1\` foot-gun (P1).** A runner started with \`ppid===1\` already (nohup, Docker as PID 1's child, launchd-spawned) would self-terminate on the first watcher tick because the condition \`current === 1\` is tautologically true. The primary path (Claude Code parent with \`pid > 1\`) is unaffected, but anyone running these runners as published packages outside Claude Code hits it. Attempted fix-then-revert on this branch — needs deeper investigation of why \`process.ppid\` polling actually fires in our happy-path smoke despite my probes showing the value is cached on Bun macOS.
- **\`server.close()\` unguarded rejection (P2).** \`shutdown()\` wraps \`dispose()\` in try/catch but \`server.close()\` is unguarded. A rejection there leaves the runner half-shut-down with \`shuttingDown=true\` so no other path can retry. Pre-existing on the shutdown path; the watcher just adds another caller.
- **Watcher re-fires every tick after detection (P3).** \`onParentDeath\` runs every \`intervalMs\` until \`process.exit\` lands. \`shutdown()\` is idempotent via the \`shuttingDown\` flag, but the lifecycle log line is outside the guard and emits N times. Fix: latch \`onParentDeath\` after first detection, or capture the watcher handle and \`clearInterval\` in \`shutdown()\`.
- **\`MCP_PARENT_CHECK_MS\` undocumented (P3).** Only the changeset and plan reference it. A short \"Environment variables\" section in each package README would help anyone tuning the watcher.
- **Parent-death log below default visibility (P3).** \`lifecycleLogger.info('Parent process gone…')\` falls below the MCP protocol sink default (\`warning\`) and the fingers-crossed flush trigger. Should log at \`warning\` so the event is visible by default.
- **Smoke negative-control variant (P3).** Current test detaches stdin via \`stdio:'ignore'\` (= \`/dev/null\`); a stdin-pipe variant would validate watcher-vs-onclose precedence in production shape.

The two safe test-only improvements from review *did* land in the second commit:
- Smoke negative control now \`await intermediate.exited\` before measuring runner liveness.
- Test comment drift fixed across the three test files.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Parent-process liveness monitoring for runners that triggers graceful shutdown when the parent disappears; polling configurable via MCP_PARENT_CHECK_MS (0 disables; very small intervals are clamped); watcher is non-blocking.

* **Bug Fixes**
  * Hardened shutdown to ensure clean exit even if server close fails.

* **Tests**
  * Added unit and smoke tests covering detection, interval handling, disabled behavior, and cross-runner scenarios.

* **Documentation**
  * Added plan and docs for preventing orphaned runner processes and test strategies.

### CODE REVIEW

File: packages/tsc-runner/mcp/index.ts [hunk 1]
```diff
@@ -1102,6 +1102,69 @@ export async function createTscServer(
...
+}): ReturnType<typeof setInterval> | undefined {
+	if (opts.intervalMs <= 0) {
+		return undefined
+	}
+	const handle = setInterval(() => {
+		const current = opts.getPpid()
+		if (current !== opts.initialPpid || current === 1) {
+			opts.onParentDeath()
+		}
+	}, opts.intervalMs)
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply packages/tsc-runner/mcp/index.ts:1168]: > Avoid immediate shutdown when `initialPpid` is already `1`.

Addressed in `ddc4255`: the watcher no longer treats `current === 1` by itself as parent death. It now shuts down when the observed PPID changes or when the original parent PID no longer exists, clears the interval before invoking shutdown, and the behavior is covered across all three runners.

File: scripts/smoke/run-smoke.ts [hunk 2]
```diff
@@ -352,6 +357,240 @@ async function runBiomeSmoke(sandboxRoot: string): Promise<void> {
...
+				}
+				try {
+					intermediate.kill('SIGKILL')
+				} catch {
+					// Ignore.
+				}
+				await intermediate.exited.catch(() => {})
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply scripts/smoke/run-smoke.ts:637]: > Consider adding explicit process-group isolation to `runOrphanDetectionSmoke`.

Addressed in `ddc4255`: the orphan smoke now launches the runner through `child_process.spawn(..., { detached: true, stdio: ["ignore", "ignore", "ignore"] })`, calls `unref()`, and cleans up the runner process group. The watcher also probes the original parent PID directly, so the detached harness still detects parent death even when Bun retains the startup `process.ppid` value. `bun run test:smoke` passes.

File: scripts/smoke/run-smoke.ts [hunk 3]
```diff
@@ -352,6 +357,240 @@ async function runBiomeSmoke(sandboxRoot: string): Promise<void> {
...
+					}
+				}
+				try {
+					intermediate.kill('SIGKILL')
+				} catch {
+					// Ignore.
+				}
+				await intermediate.exited.catch(() => {})
+			}
+		}
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply scripts/smoke/run-smoke.ts:608]: > Negative-control failure indicates Bun’s `stdio: ["ignore", ...]` is not detaching stdin as assumed.

Addressed in `ddc4255`: the smoke harness now uses explicit detached process-group spawning with ignored stdio and group cleanup. The runner watcher was also hardened to detect a dead original parent PID instead of relying only on `process.ppid`, which can be stale in Bun. Verified with `bun run test:smoke`.

**Comment by nathanvale** [comment]
> Make parent-death watcher one-shot to prevent repeated callbacks.

Addressed in `ddc4255`: the parent-liveness watcher now clears its interval before invoking `onParentDeath`, and the one-shot behavior is covered in the tsc, bun, and biome runner tests.

File: packages/biome-runner/mcp/index.ts [hunk 4]
```diff
@@ -1256,6 +1339,22 @@ export async function startBiomeServer(): Promise<void> {
...
+		getPpid: () => process.ppid,
+		isParentAlive: isPidAlive,
+		intervalMs,
+		onParentDeath: () => {
+			lifecycleLogger.info('Parent process gone, shutting down', {
+				initialPpid,
+				currentPpid: process.ppid,
+			})
+			void shutdown()
+		},
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply packages/biome-runner/mcp/index.ts]: > Make the parent-death path exit even if `server.close()` fails.

Addressed in `e88c809`: `shutdown()` now wraps `server.close()` in `try/catch` and calls `process.exit(0)` in `finally`, so once shutdown starts the biome runner exits even if server close rejects. Applied the same hardening across all three runners and revalidated with Biome, TypeScript, full Bun tests, and `bun run test:smoke`.

File: packages/tsc-runner/mcp/index.ts [hunk 5]
```diff
@@ -1144,6 +1227,22 @@ export async function startTscServer(): Promise<void> {
...
+		getPpid: () => process.ppid,
+		isParentAlive: isPidAlive,
+		intervalMs,
+		onParentDeath: () => {
+			lifecycleLogger.info('Parent process gone, shutting down', {
+				initialPpid,
+				currentPpid: process.ppid,
+			})
+			void shutdown()
+		},
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply packages/tsc-runner/mcp/index.ts]: > A `server.close()` rejection still defeats orphan cleanup here.

Addressed in `e88c809`: `shutdown()` now catches `server.close()` failures, logs them, and exits in `finally`, guaranteeing the tsc runner terminates after parent-death shutdown begins. The same pattern was applied to bun and biome for parity. Validation passed: Biome, TypeScript, full Bun tests, and `bun run test:smoke`.

File: packages/biome-runner/mcp/index.ts [hunk 6]
```diff
@@ -1256,6 +1346,22 @@ export async function startBiomeServer(): Promise<void> {
...
+	})
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply packages/biome-runner/mcp/index.ts]: > **Move `initialPpid` capture to the top of `startBiomeServer()`.**
>
> The parent process PID must be captured before async initialization. If the parent dies while awaiting `createBiomeServer()` (line 1306), `process.ppid` gets reparented to init (PID 1).

Addressed in 09073ea: `const initialPpid = process.ppid` is now the first line in `startBiomeServer()`, before any `await`. Same fix applied to bun-runner and tsc-runner for consistency. Strengthened the `.unref()` test to spy on the Timer prototype rather than rely on test-exit timing.

File: packages/bun-runner/mcp/index.ts [hunk 7]
```diff
@@ -1177,6 +1267,22 @@ export async function startBunServer(): Promise<void> {
...
+		isParentAlive: isPidAlive,
+		intervalMs,
+		onParentDeath: () => {
+			lifecycleLogger.info('Parent process gone, shutting down', {
+				initialPpid,
+				currentPpid: process.ppid,
+			})
+			void shutdown()
+		},
+	})
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply packages/bun-runner/mcp/index.ts]: > **Move `initialPpid` capture to function entry, before any async operations.**
>
> The parent PID is currently captured at line 1271, after `await createBunServer()` (line 1227) and `await server.connect(transport)` (line 1260). [...] The same issue exists in `packages/biome-runner/mcp/index.ts` at line 1350.

Addressed in 09073ea: `const initialPpid = process.ppid` is now the first line in `startBunServer()` (and `startBiomeServer()` and `startTscServer()`), before any `await`. Strengthened the `.unref()` test in all three runners to spy on the Timer prototype rather than rely on test-exit timing.
